modules/aidemBidAdapter.js

```diff
@@ -0,0 +1,691 @@
+  }
+  if (isBoolean(coppa)) {
+    deepSetValue(regs, 'coppa', !!coppa);
+  }
+
+  return regs;
+}
+
+/**
+ * Returns bidder request page url.
+ *
+ * @param {Object} bidderRequest
+ * @return {string}
+ */
+function getPageUrl(bidderRequest) {
+  return bidderRequest?.refererInfo?.page
+}
+
+function buildWonBidNotice(bid) {
+  const winNotice = {
+    adId: bid.adId,
+    adUnitCode: bid.adUnitCode,
+    creativeId: bid.creativeId,
+    cpm: bid.cpm,
+    netRevenue: bid.netRevenue,
+    adserverTargeting: bid.adserverTargeting,
+    auctionId: bid.auctionId,
+    currency: bid.currency,
+    mediaType: bid.mediaType,
+    size: bid.size,
+    width: bid.width,
+    height: bid.height,
+    status: bid.status,
+    transactionId: bid.transactionId,
+    ttl: bid.ttl,
+    requestTimestamp: bid.requestTimestamp,
+    responseTimestamp: bid.responseTimestamp,
+    metrics: {}
+  }
+
+  if (bid.metrics && typeof bid.metrics.getMetrics === 'function') {
+    const metrics = bid.metrics.getMetrics()
+    deepSetValue(winNotice.metrics, 'requestBids.validate', metrics['requestBids.validate'])
+    deepSetValue(winNotice.metrics, 'requestBids.makeRequests', metrics['requestBids.makeRequests'])
+    deepSetValue(winNotice.metrics, 'requestBids.total', metrics['requestBids.total'])
+    deepSetValue(winNotice.metrics, 'requestBids.callBids', metrics['requestBids.callBids'])
+    deepSetValue(winNotice.metrics, 'adapter.client.validate', metrics['adapter.client.validate'])
+    deepSetValue(winNotice.metrics, 'adapter.client.buildRequests', metrics['adapter.client.buildRequests'])
+    deepSetValue(winNotice.metrics, 'adapter.client.total', metrics['adapter.client.total'])
+    deepSetValue(winNotice.metrics, 'adapter.client.net', metrics['adapter.client.net'])
+    deepSetValue(winNotice.metrics, 'adapter.client.interpretResponse', metrics['adapter.client.interpretResponse'])
+    deepSetValue(winNotice.metrics, 'addBidResponse.validate', metrics['addBidResponse.validate'])
+    deepSetValue(winNotice.metrics, 'addBidResponse.total', metrics['addBidResponse.total'])
+    deepSetValue(winNotice.metrics, 'render.pending', metrics['render.pending'])
+    deepSetValue(winNotice.metrics, 'render.e2e', metrics['render.e2e'])
+  }
+
+  return winNotice
+}
+
+// Called for every bid that has timed out
+function buildTimeoutNotice(prebidTimeoutBids) {
+  const timeoutNotice = {
+    bids: []
+  }
+  _each(prebidTimeoutBids, function (bid) {
+    const notice = {
+      adUnitCode: bid.adUnitCode,
+      auctionId: bid.auctionId,
+      bidId: bid.bidId,
+      bidderRequestId: bid.bidderRequestId,
+      transactionId: bid.transactionId,
+      mediaTypes: bid.mediaTypes,
+      metrics: { },
+      timeout: bid.timeout,
+    }
+
+    if (bid.metrics && typeof bid.metrics.getMetrics === 'function') {
+      const metrics = bid.metrics.getMetrics()
+      deepSetValue(notice.metrics, 'requestBids.validate', metrics['requestBids.validate'])
```

Comment:
I am not sure it's appropriate for a bid adapter to collect this sort of information (@patmmccann ?), but if it is, I suggest saving a few bytes with `['requestBids.validate', 'requestBids.makeRequests', ....].forEach((n) => deepSetValue(notice.metrics, n, metrics[n]))`

Reply:
@dgirardi removed metrics collection.